run-clang-format.py

```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/env python
```

Comment:
What's the benefit of this script over the standard CLI? pre-commit can handle the parallel parts if that's relevant. In general, I'd encourage the use of pre-commit over manual management of the hooks. I think freud/signac/etc. style checks have gotten a lot easier to maintain since adopting pre-commit. It's extremely automation-friendly.

Reply:
clang-format does not have built in support for recursive runs or printing diffs. This script (which I copied from somewhere) does. I have no experience with `pre-commit`, so converting the scripts to that would require a large learning curve.

I do not plan to manually manage git hooks, just use the CI to make the checks on PRs and educate users (in `style.rst`) how to run the formatters when needed. Aside from the time to learn a new tool, I have two concerns with pre-commit hooks. 1) Formatting takes a long time - I don't want to wait minutes for checks to run every time I run `git commit`. 2) I've had issues in the past when running merge or rebase operations when intermediate commits fail the hooks.

Reply:
I see, that could be helpful for clang-format. If it's useful, we used to use this snippet for freud:
```bash
        for f in cpp/**/*; do
          if ! diff $f <(clang-format-6.0 -style=file $f) > /dev/null; then
            echo "$f does not match clang-format style."
            diff --color $f <(clang-format-6.0 -style=file $f) || true
          fi
        done
```

Pre-commit hooks only run on changed files and execute in parallel by default, so your concern about time to run `git commit` might not be a problem. I never felt any significant slowdown (maybe 1-2 seconds) when using [cuDF's pre-commit hooks](https://github.com/rapidsai/cudf/blob/branch-0.20/.pre-commit-config.yaml) on a similarly large C++ / Python code base. One reason I like pre-commit is that users can opt in for local runs but don't _have_ to, you can always rely on CI to catch problems. However, there's not an easy way to execute these checks with a git hook for developers that prefer that workflow. I leave this up to you, just wanted to suggest it since I've been so happy with the tool.

Reply:
Another aspect not already mentioned is versioning: developers using yapf / flake8 locally may not have the same versions as the container, and that can be quite troublesome. Using pre-commit allows much finer control over hook versions, and can be automatically updated for all developers using pre-commit if the configuration changes.

Reply:
Feel fee to contribute an equivalent `pre-commit` config to this PR. That would be faster than me attempting to learn how the system works.

Note that the docker image I'm using has the following flake8 plugins installed:
* pep8-naming
* flake8-docstrings
* flake8-rst-docstrings

Reply:
Also, the clang-format version needs to be relatively recent to support the config I use (>= 10).

Reply:
This can be removed, resolved by pre-commit #983.